### DISCUSSION TITLE
Need help with implementing graves attention layer

### DISCUSSION THREAD

**Discussion by JRMeyer** [coqui-ai/TTS#270] [Archived]
<USER>>>> justin_tian<br/> <DATE> [September 10, 2020, 3:49pm]<br/> <br/> I'm trying to implement graves (GMM) attention based on Mozilla TTS<br/> repo. Here is a link with brief discussions about the implementation by<br/> the repo holder. Code below is my implementation to fit<br/> flowtron(<https://github.com/NVIDIA/flowtron>).<br/> <br/> The author of flowtron make some changes to Tacotron. They feed all<br/> encoded mel frames once so the dimension of queries, keys, and values<br/> are<br/> <br/> queries: T_mel *slash* curl-run-all.sh discourse.mozilla.org html-to-markdown.sh ordered-posts ordered-posts~ TTS.cdx tts.commands tts-emails.txt TTS.pages tts-telegram.txt TTS.warc.gz B *slash* curl-run-all.sh discourse.mozilla.org html-to-markdown.sh ordered-posts ordered-posts~ TTS.cdx tts.commands tts-emails.txt TTS.pages tts-telegram.txt TTS.warc.gz attn_hidden_dim *slash* <br/> keys, values: T_text *slash* curl-run-all.sh discourse.mozilla.org html-to-markdown.sh ordered-posts ordered-posts~ TTS.cdx tts.commands tts-emails.txt TTS.pages tts-telegram.txt TTS.warc.gz B *slash* curl-run-all.sh discourse.mozilla.org html-to-markdown.sh ordered-posts ordered-posts~ TTS.cdx tts.commands tts-emails.txt TTS.pages tts-telegram.txt TTS.warc.gz text_embedding_dim<br/> <br/> When I train it, it just doesn't work well. Only the first frame<br/> alignment is close to maximum value and other frame attn. scores are<br/> really low. Also, the converging speed is slow (comparing to an additive<br/> attn). Can someone help me out to see which part of the code needs a fix<br/> (possibly the mu_t part)?<br/> <br/> class GravesAttention(torch.nn.Module):<br/> <br/> def __init__(self, n_mel_channels=80, n_speaker_dim=128,<br/> n_text_channels=512, n_att_channels=256, K=4):<br/> super(GravesAttention, self).__init__()<br/> ## K is number of gaussian component<br/> self.K = K<br/> self._mask_value = 1e-8<br/> self.eps = 1e-5<br/> self.J = None<br/> self.N_a = nn.Sequential(<br/> nn.Linear(n_mel_channels, n_mel_channels, bias=True),<br/> nn.ReLU(),<br/> nn.Linear(n_mel_channels, 3 curl-run-all.sh discourse.mozilla.org html-to-markdown.sh ordered-posts ordered-posts~ TTS.cdx tts.commands tts-emails.txt TTS.pages tts-telegram.txt TTS.warc.gz K, bias=True)<br/> )<br/> self.key = LinearNorm(n_text_channels + n_speaker_dim,<br/> n_att_channels, bias=False, w_init_gain='tanh')<br/> self.value = LinearNorm(n_text_channels + n_speaker_dim,<br/> n_att_channels, bias=False,<br/> w_init_gain='tanh')<br/> self.init_layers()<br/> <br/> def init_layers(self):<br/> torch.nn.init.constant_(self.N_a[2].bias[(2 curl-run-all.sh discourse.mozilla.org html-to-markdown.sh ordered-posts ordered-posts~ TTS.cdx tts.commands tts-emails.txt TTS.pages tts-telegram.txt TTS.warc.gz self.K):(3 curl-run-all.sh discourse.mozilla.org html-to-markdown.sh ordered-posts ordered-posts~ TTS.cdx tts.commands tts-emails.txt TTS.pages tts-telegram.txt TTS.warc.gz self.K)], 1.) # bias mean<br/> torch.nn.init.constant_(self.N_a[2].bias[self.K:(2 curl-run-all.sh discourse.mozilla.org html-to-markdown.sh ordered-posts ordered-posts~ TTS.cdx tts.commands tts-emails.txt TTS.pages tts-telegram.txt TTS.warc.gz self.K)], 10) # bias std<br/> <br/> def init_states(self, inputs):<br/> if self.J is None or inputs.shape[0] + 1 > self.J.shape[-1]:<br/> self.J = torch.arange(0, inputs.shape[0] + 2.0).to(inputs.device) + 0.5<br/> <br/> def forward(self, queries, keys, values, mask=None, attn=None):<br/> <br/> self.init_states(keys) ##initialize self.J<br/> if attn is None:<br/> keys = self.key(keys).transpose(0, 1) # B x in_lens x n_attn_channels<br/> <br/> values = self.value(values) if hasattr(self, 'value') else values<br/> values = values.transpose(0, 1) # B x in_lens x n_attn_channels<br/> <br/> gbk_t = self.N_a(queries).transpose(0, 1) # B x T x 3K<br/> <br/> gbk_t = gbk_t.view(gbk_t.size(0), gbk_t.size(1), -1, self.K)<br/> # each B x T x K<br/> g_t = gbk_t[:, :, 0, :]<br/> b_t = gbk_t[:, :, 1, :]<br/> k_t = gbk_t[:, :, 2, :]<br/> <br/> g_t = torch.nn.functional.dropout(g_t, p=0.5, training=self.training)<br/> <br/> sig_t = torch.nn.functional.softplus(b_t) + self.eps<br/> k_t = torch.nn.functional.softplus(k_t)<br/> mu_t = torch.cumsum(k_t, dim=1) ## mu_t = mu_(t-1) + k_t, mu_0 = 0<br/> g_t = torch.softmax(g_t, dim=-1) + self.eps<br/> j = self.J[:values.size(1) + 1]<br/> <br/> <br/> phi_t = g_t.unsqueeze(-1) curl-run-all.sh discourse.mozilla.org html-to-markdown.sh ordered-posts ordered-posts~ TTS.cdx tts.commands tts-emails.txt TTS.pages tts-telegram.txt TTS.warc.gz (1 / (1 + torch.sigmoid((mu_t.unsqueeze(-1) - j) /<br/> sig_t.unsqueeze(-1))))<br/> <br/> alpha_t = torch.sum(phi_t, 2) ## sum over attn heads<br/> alpha_t = alpha_t[:, :, 1:] - alpha_t[:, :, :-1]<br/> <br/> alpha_t[alpha_t == 0] = 1e-8<br/> <br/> if mask is not None:<br/> alpha_t.data.masked_fill_(mask.transpose(1, 2), self._mask_value)<br/> else:<br/> values = self.value(values)<br/> values = values.transpose(0, 1)<br/> print('with_dropout flows2 max, min in alpha_t {} {}'.format(torch.max(alpha_t), torch.min(alpha_t)))<br/> output = torch.bmm(alpha_t, values)<br/> output = output.transpose(1, 2)<br/> <br/> return output, alpha_t<br/> <br/> [This is an archived TTS discussion thread from discourse.mozilla.org/t/need-help-with-implementing-graves-attention-layer]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [September 11, 2020, 10:18am]<br/> <br/> I am confused. D you implement it for mozilla TTS or flowtron?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> justin_tian<br/> <DATE> [September 11, 2020, 10:49am]<br/> <br/> Hi erogol, thanks for the reply. I'm doing it for flowtron. Training on<br/> flowtron really drives me crazy. I just don't know if the problem is<br/> caused by tricky training steps of flowtron or it's my implementation of<br/> graves attn. for flowtron is wrong.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> justin_tian<br/> <DATE> [September 14, 2020, 3:01am]<br/> <br/> Okay, I think that I get the bitter answer to my question. It looks like<br/> that any dynamic attention mechanism like graves or forward attention<br/> cannot work with flow based models. Flowtron is using a memory-less<br/> attention layer which uses only the local query and key vectors to<br/> calculate alignment score. In any location sensitive attention layer,<br/> the score at position j depends on alignment results at j-1, j-2 ... It<br/> looks like the only improvement I can try now on flowtron is adding a<br/> bi-directional attention flow. If anyone knows anything other than BiDAF<br/> that works with flow based model, please let me know.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [September 14, 2020, 1:27pm]<br/> <br/> I don't think there is such limitation since you can still process<br/> temporal dimension iteratively for GMM and Location sensitive attention.<br/> But I may need to think about it more.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> justin_tian<br/> <DATE> [September 14, 2020, 1:45pm]<br/> <br/> I'm thinking about to let the N_a layer predict mu_t directly with<br/> sigmoid activation times the length of text input. I will test this<br/> today and let you know about the results.<br/> <br/> [Archived Post]<br/>
